xdsl/dialects/riscv.py

```diff
@@ -48,19 +49,67 @@
 from xdsl.utils.hints import isa
 
 
-@dataclass(frozen=True)
-class Register:
+class RISCVRegisterType(Data[str], TypeAttribute, ABC):
     """
-    A RISC-V register.
+    A RISC-V register type.
     """
 
-    name: str | None = field(default=None)
-    """The register name. Should be one of `ABI_INDEX_BY_NAME` or `None`"""
+    _unallocated: ClassVar[Self | None] = None
+
+    @classmethod
+    def unallocated(cls) -> Self:
+        if cls._unallocated is None:
+            cls._unallocated = cls("")
+        return cls._unallocated
+
+    @property
+    def register_name(self) -> str:
+        """Returns name if allocated, raises ValueError if not"""
+        if not self.is_allocated:
+            raise ValueError("Cannot get name for unallocated register")
```

Comment:
I'm not sure why we need the logic here. Can't we return empty string?

Reply:
data is the empty string, this function guarantees that the register is allocated

Reply:
I feel that is a valid question. In theory, we could return an empty string for unallocated registers. The current code base does not call `register_name` for unallocated registers, so that codepath would be untested. I can see a valid argument, that we want for now to stay with an assertion as this is the only use case we consider. We can always change this if we feel other use-cases would benefit from register_name defaulting to `""` for unallocated registers. However, I can see this being confusing, so maybe lets stay with the current assertion?